File: src/main/java/com/aditi/moneymanager/model/BudgetModel.java
package com.aditi.moneymanager.model;

import java.sql.Date;
import java.util.List;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;
import lombok.Data;

@Data
@Entity
@Table(name = "Budget")
public class BudgetModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Date startDate;

    @Column(nullable = false)
    private Date endDate;

    @Column(nullable = false)
    private double budget;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private UserModel user;

    @ManyToMany
    @JoinTable(
        name = "budget_categories",
        joinColumns = @JoinColumn(name = "budget_id"),
        inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private List<CategoryModel> categories;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BudgetPeriod period;

    @CreationTimestamp
    @Column(updatable = false)
    private Date createdAt;

    @UpdateTimestamp
    private Date updatedAt;

    @Column(length = 255)
    private String description;

    public enum BudgetPeriod {
        DAILY,
        WEEKLY,
        MONTHLY,
        CUSTOM
    }
}

File: models/budgetModel.js
import { DataTypes } from "sequelize";
import sequelize from "../config/database.js";
import User from "./user.model.js";
import Category from "./category.model.js";

const Budget = sequelize.define("Budget", {
  id: {
    type: DataTypes.BIGINT,
    autoIncrement: true,
    primaryKey: true,
  },
  startDate: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  endDate: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  budget: {
    type: DataTypes.DOUBLE,
    allowNull: false,
  },
  period: {
    type: DataTypes.ENUM("DAILY", "WEEKLY", "MONTHLY", "CUSTOM"),
    allowNull: false,
  },
  description: {
    type: DataTypes.STRING(255),
  },
  createdAt: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW,
  },
  updatedAt: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW,
  },
});

Budget.belongsTo(User, { foreignKey: "user_id" });
Budget.belongsToMany(Category, { through: "budget_categories" });

export default Budget;